### PR TITLE
Create 'expire' and 'repack' verbs for git-multi-pack-index

### PR DESCRIPTION
**Pull request by derrickstolee**
The multi-pack-index provides a fast way to find an object among a large list of pack-files. It stores a single pack-reference for each object id, so duplicate objects are ignored. Among a list of pack-files storing the same object, the most-recently modified one is used.

Create new subcommands for the multi-pack-index builtin.

* 'git multi-pack-index expire': If we have a pack-file indexed by the multi-pack-index, but all objects in that pack are duplicated in more-recently modified packs, then delete that pack (and any others like it). Delete the reference to that pack in the multi-pack-index.

* 'git multi-pack-index repack --batch-size=<size>': Starting from the oldest pack-files covered by the multi-pack-index, find those whose "expected size" is below the batch size until we have a collection of packs whose expected sizes add up to the batch size. We compute the expected size by multiplying the number of referenced objects by the pack-size and dividing by the total number of objects in the pack. If the batch-size is zero, then select all packs. Create a new pack containing all objects that the multi-pack-index references to those packs.

This allows us to create a new pattern for repacking objects: run 'repack'. After enough time has passed that all Git commands that started before the last 'repack' are finished, run 'expire' again. This approach has some advantages over the existing "repack everything" model:

1. Incremental. We can repack a small batch of objects at a time, instead of repacking all reachable objects. We can also limit ourselves to the objects that do not appear in newer pack-files.

2. Highly Available. By adding a new pack-file (and not deleting the old pack-files) we do not interrupt concurrent Git commands, and do not suffer performance degradation. By expiring only pack-files that have no referenced objects, we know that Git commands that are doing normal object lookups* will not be interrupted.

* Note: if someone concurrently runs a Git command that uses get_all_packs(), then that command could try to read the pack-files and pack-indexes that we are deleting during an expire command. Such commands are usually related to object maintenance (i.e. fsck, gc, pack-objects) or are related to less-often-used features (i.e. fast-import, http-backend, server-info).

We **are using** this approach in VFS for Git to do background maintenance of the "shared object cache" which is a Git alternate directory filled with packfiles containing commits and trees. We currently download pack-files on an hourly basis to keep up-to-date with the central server. The cache servers supply packs on an hourly and daily basis, so most of the hourly packs become useless after a new daily pack is downloaded. The 'expire' command would clear out most of those packs, but many will still remain with fewer than 100 objects remaining. The 'repack' command (with a batch size of 1-3gb, probably) can condense the remaining packs in commands that run for 1-3 min at a time. Since the daily packs range from 100-250mb, we will also combine and condense those packs.

Updates in V2:

* Added a method, `unlink_pack_path()` to remove packfiles, but with the additional check for a `.keep` file. This borrows logic from `builtin/repack.c`.

* Modified documentation and commit messages to replace 'verb' with 'subcommand'. Simplified the documentation. (I left 'verbs' in the title of the cover letter for consistency.)

Updates in V3:

* There was a bug in the expire logic when simultaneously removing packs and adding uncovered packs, specifically around the pack permutation. This was hard to see during review because I was using the 'pack_perm' array for multiple purposes. First, I was reducing its length, and then I was adding to it and resorting. In V3, I significantly overhauled the logic here, which required some extra commits before implementing 'expire'. The final commit includes a test that would cover this case.

Updates in V4:

* More 'verb' and 'command' instances replaced with 'subcommand'. I grepped the patch to check these should be fixed everywhere.

* Update the tests to check .keep files (in last patch).

* Modify the tests to show the terminating condition of --batch-size when there are three packs that fit under the size, but the first two are large enough to stop adding packs. This required rearranging the packs slightly to get different sizes than we had before. Also, I added 'touch -t' to set the modified times so we can fix the order in which the packs are selected.

* Added a comment about the purpose of pack_perm.

Updates in V5:

* Fixed the error in PATCH 7 due to a missing line that existed in PATCH 8. Thanks, Josh Steadmon!

* The 'repack' subcommand now computes the "expected size" of a pack instead of relying on the total size of the pack. This is actually really important to the way VFS for Git uses prefetch packs, and some packs are not being repacked because the pack size is larger than the batch size, but really there are only a few referenced objects.

* The 'repack' subcommand now allows a batch size of zero to mean "create one pack containing all objects in the multi-pack-index". A new commit adds a test that hits the boundary cases here, but follows the 'expire' subcommand so we can show that cycle of repack-then-expire to safely replace the packs.

Junio: It appears that there are some conflicts with the trace2 changes in master. These are not new to the updates in this version. I'm assuming you've resolved these conflicts and replaying that resolution will work for you. If not, then here are my recommendations:

In midx.c, it's just a conflict on the last `#include` statement, so any order is fine.

In builtin/multi-pack-index.c:

```
<<<<<<< HEAD
        if (!strcmp(argv[0], "repack"))
                return midx_repack(opts.object_dir, (size_t)opts.batch_size);
        if (opts.batch_size)
                die(_("--batch-size option is only for 'repack' subcommand"));
=======
        trace2_cmd_mode(argv[0]);
>>>>>>> origin/master
```

Place the trace2_cmd_mode(argv[0]) above the strcmp.

Thanks,
-Stolee

Cc: sbeller@google.com, peff@peff.net, jrnieder@gmail.com, avarab@gmail.com, jonathantanmy@google.com

### CODE REVIEW

File: t/t5319-multi-pack-index.sh [hunk 1]
```diff
@@ -348,4 +348,92 @@ test_expect_success 'verify incorrect 64-bit offset' '
...
+		cd dup &&
+		ls .git/objects/pack >expect &&
+		MINSIZE=$(ls -l .git/objects/pack/*pack | awk "{print \$5;}" | sort -n | head -n 1) &&
+		git multi-pack-index repack --batch-size=$MINSIZE &&
+		ls .git/objects/pack >actual &&
+		test_cmp expect actual
+	)
+'
+
+test_expect_success 'repack creates a new pack' '
```
Comment (derrickstolee) [inline t/t5319-multi-pack-index.sh]: explicitly test 'expire' after 'repack' to show that certain packs are removed.

File: t/t5319-multi-pack-index.sh [hunk 2]
```diff
@@ -348,4 +348,102 @@ test_expect_success 'verify incorrect 64-bit offset' '
...
+		test_line_count = 5 idx-list &&
+		test-tool read-midx .git/objects | grep idx >midx-list &&
+		test_line_count = 5 midx-list
+	)
+'
+
+test_expect_success 'expire removes repacked packs' '
+	(
+		cd dup &&
+		ls --sort=size .git/objects/pack/*pack | head -n 3 >expect &&
```
Comment (dscho) [inline t/t5319-multi-pack-index.sh]: Please note that `ls --sort=size` seems not to be portable. How is it done elsewhere in the test suite?

Comment (derrickstolee) [reply t/t5319-multi-pack-index.sh]: Yeah, I noticed this, too. The POSIX way is `-S`.

**Comment by dscho** [comment]
Please do not forget to re-submit before long, lest v4 goes into `next`.

**Comment by dscho** [comment]
@derrickstolee maybe close this here PR?

**Comment by dscho** [comment]
@derrickstolee maybe close this here PR?
